lightning-block-sync/src/http_clients.rs

```diff
@@ -0,0 +1,1080 @@
+
+use bitcoin::blockdata::block::{Block, BlockHeader};
+use bitcoin::consensus::encode;
+use bitcoin::hash_types::{BlockHash, TxMerkleNode};
+use bitcoin::hashes::hex::{ToHex, FromHex};
+
+use serde_derive::Deserialize;
+
+use serde_json;
+
+use std::cmp;
+use std::convert::TryFrom;
+use std::convert::TryInto;
+use std::io::Write;
+use std::net::ToSocketAddrs;
+use std::time::Duration;
+
+#[cfg(feature = "rpc-client")]
+use base64;
+#[cfg(feature = "rpc-client")]
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[cfg(feature = "tokio")]
+use tokio::io::AsyncReadExt;
+#[cfg(feature = "tokio")]
+use tokio::net::TcpStream;
+
+#[cfg(not(feature = "tokio"))]
+use std::io::Read;
+#[cfg(not(feature = "tokio"))]
+use std::net::TcpStream;
+
+/// Maximum HTTP response size in bytes.
+const MAX_HTTP_RESPONSE_LEN: usize = 4_000_000;
+
+/// Client for making HTTP requests.
+struct HttpClient {
+	stream: TcpStream,
+}
+
+impl HttpClient {
+	/// Opens a connection to an HTTP endpoint.
+	fn connect<E: ToSocketAddrs>(endpoint: E) -> std::io::Result<Self> {
+		let address = match endpoint.to_socket_addrs()?.next() {
+			None => {
+				return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "could not resolve to any addresses"));
+			},
+			Some(address) => address,
+		};
+		let stream = std::net::TcpStream::connect_timeout(&address, Duration::from_secs(1))?;
+		stream.set_read_timeout(Some(Duration::from_secs(2)))?;
+		stream.set_write_timeout(Some(Duration::from_secs(1)))?;
+
+		#[cfg(feature = "tokio")]
+		let stream = TcpStream::from_std(stream)?;
+
+		Ok(Self { stream })
+	}
+
+	/// Sends a `GET` request for a resource identified by `uri` at the `host`.
+	async fn get<F>(&mut self, uri: &str, host: &str) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		write!(self.stream,
+			"GET {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 \r\n", uri, host)?;
+
+		let bytes = read_http_resp(&self.stream, MAX_HTTP_RESPONSE_LEN).await?;
+		F::try_from(bytes)
+	}
+
+	/// Sends a `POST` request for a resource identified by `uri` at the `host` using the given HTTP
+	/// authentication credentials.
+	///
+	/// The request body consists of the provided JSON `content`. Returns the response body in `F`
+	/// format.
+	async fn post<F>(&mut self, uri: &str, host: &str, auth: &str, content: serde_json::Value) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		let content = content.to_string();
+		write!(self.stream,
+			"POST {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Authorization: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 Content-Type: application/json\r\n\
+			 Content-Length: {}\r\n\
+			 \r\n\
+			 {}", uri, host, auth, content.len(), content)?;
+
+		let bytes = read_http_resp(&self.stream, MAX_HTTP_RESPONSE_LEN).await?;
+		F::try_from(bytes)
+	}
+}
+
+async fn read_http_resp(mut socket: &TcpStream, max_resp: usize) -> std::io::Result<Vec<u8>> {
```

Comment:
Yeah, I tend to agree thought didn't get around into investigating this fully. Thanks for pointing out some of the alternatives. Will spend a little time looking into this.

Reply:
in case this does stay with HTTP, why not use one of the HTTP client crates that are already out there?

Reply:
This is definitely a consideration as per https://github.com/rust-bitcoin/rust-lightning/pull/763#discussion_r534527616. From offline discussions, the URI parsing can be removed entirely by simply having the user of `HttpEndpoint` pass the components parts as parameters instead of the URI string. I took a stab at this yesterday with good results. But I still plan on investigating whether there is a suitable, off-the-shelf replacement for `read_http_resp`.

Reply:
Sorry, I didn't read Matt's comment carefully enough.

Reply:
Maybe could steal the `hyper` response reader - https://github.com/hyperium/hyper/blob/master/src/proto/h1/decode.rs